api/buckets/service.go

```diff
@@ -290,126 +289,194 @@ func (s *Service) newDirFromExistingPath(ctx context.Context, pth path.Path, key
-			name: name,
-			node: dag.NewRawNode(data),
-		}
-	case *dag.ProtoNode:
-		var isLeaf bool
-		var leaves []*treeNode
-		// Follow each link, wrapping up and encrypting files and dirs
-	loop:
-		for _, l := range root.Links() {
-			if l.Name == "" {
-				isLeaf = true
-				break loop // We have discovered a raw node wrapper
-			}
-			ln, err := l.GetNode(ctx, ds)
-			if err != nil {
-				return nil, err
-			}
-			ltn, err := s.encryptTree(ctx, ds, tree, ln, l.Name, key, nil)
-			if err != nil {
-				return nil, err
-			}
-			leaves = append(leaves, ltn)
-		}
-		if !isLeaf {
-			// We're done with these links, wrap up the new dir
-			dir := unixfs.EmptyDirNode()
-			dir.SetCidBuilder(dag.V1CidPrefix())
-			for _, l := range leaves {
-				if err := dir.AddNodeLink(l.name, l.node); err != nil {
-					return nil, err
+// This method returns a map of all nodes keyed by their _original_ plaintext cid.
+func (s *Service) encryptDag(ctx context.Context, ds ipld.DAGService, root ipld.Node, key []byte, add *namedNode) (map[cid.Cid]*namedNode, error) {
+	// Step 1: Create a preordered list of joint and leaf nodes
+	var stack, joints, leaves []*namedNode
+	stack = append(stack, &namedNode{node: root})
+	var cur *namedNode
+
+	for len(stack) > 0 {
+		n := len(stack) - 1
+		cur = stack[n]
+		stack = stack[:n]
+	types:
+		switch cur.node.(type) {
+		case *dag.RawNode:
+			leaves = append(leaves, cur)
+		case *dag.ProtoNode:
+			// Add links to the stack
+			for _, l := range cur.node.Links() {
+				if l.Name == "" {
+					leaves = append(leaves, cur)
+					break types // We have discovered a raw node wrapper
 				}
-			}
-			if add != nil {
-				if err := dir.AddNodeLink(add.name, add.node); err != nil {
+				ln, err := l.GetNode(ctx, ds)
+				if err != nil {
 					return nil, err
 				}
-				tree[add.node.Cid()] = add
+				stack = append(stack, &namedNode{name: l.Name, node: ln})
 			}
-			cn, err := encryptNode(dir, key)
-			if err != nil {
-				return nil, err
+			joints = append(joints, cur)
+		default:
+			return nil, errInvalidNodeType
+		}
+	}
+
+	// Step 2: Encrypt all leaf nodes in parallel
+	nmap := newNamedNodes()
+	var wg sync.WaitGroup
+	errs := make(chan error)
+	wgDone := make(chan bool)
+	for _, l := range leaves {
+		wg.Add(1)
+		go func(l *namedNode) {
+			defer wg.Done()
+			var cn ipld.Node
+			switch l.node.(type) {
+			case *dag.RawNode:
+				data, err := encryptData(l.node.RawData(), key)
+				if err != nil {
+					errs <- err
+					return
+				}
+				cn = dag.NewRawNode(data)
+			case *dag.ProtoNode:
+				var err error
+				cn, err = s.encryptFileNode(ctx, l.node, key)
+				if err != nil {
+					errs <- err
+					return
+				}
 			}
-			tn = &treeNode{
-				name: name,
+			nmap.Store(l.node.Cid(), &namedNode{
+				name: l.name,
 				node: cn,
-			}
-			break
-		}
+			})
+		}(l)
+	}
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-errs:
+		close(errs)
+		return nil, err
```

Comment:
I see some possible problems here, so speaking loud to double-check:

In a weird scenario, `wgDone` could be closed and still someone used to `errs`. Maybe most probably if the last encryption might fail since both `<-wgDone` and `<-errs` are _ok_ for the select case. So possibly, the select might enter to `wgDone` when some encryption failed. This might blow up later in reconstructing the DAG anyway since it won't find the encrypted node that failed. So maybe, can be ignored... but something tricky.

Another, possibly panicking, problem is that `close(errs)` is done. That means if another spun goroutine in the background also pushes to `errs`, it will panic.

Furthermore, since this returns early (`return nil, err`), the other spun goroutines will still be doing work (even if they don't error, which is a bit _useless_).

A solution might be allowing the goroutines to receive a ctx and just check if `ctx.Err() != nil -> return ` to fail fast. And use [errgroup](https://godoc.org/golang.org/x/sync/errgroup), which basically runs a set of goroutines and auto-cancels the ctx as soon as the first errors and returns that error. Which is pretty similar to the intention here.

Reply:
Nice, `errgroup` sounds much cleaner. Changed.

Reply:
Nice!